### PR TITLE
Add API to create armature prototypes and join them together

### PR DESCRIPTION
**Pull request by davepagurek**
This addresses https://github.com/calder-gl/calder/issues/35.

This exposes an API of named points so that nodes can be attached to each other. For example, from the example file:
```typescript
nextPiece.point('base').stickTo(top.point('tip'));
```

To set up these named points, you can create a named armature prototype using `Armature.define`, which you can then make multiple instances of.
```typescript
const bone = Armature.define((root: Node) => {
    root.createPoint('base', vec3.fromValues(0, 0, 0));
    root.createPoint('tip', vec3.fromValues(0, 1, 0));
});
const parent = bone();
```

## Other changes
- There were small bugs in the definitions of bone and sphere geometry that I fixed
- I randomized sphere colours and animated their rotation to check that that was working
- Bones are still shadeless but are coloured based on their x, y, and z coordinates so you can get a sense of their orientation
screen shot 2018-05-13 at 2 49 42 pm

## Issues I want opinions on
- `Armature.define` returns basically a node generator, which you use kind of like a class. Should I be doing something like `makeBone = Armature.define(...)` or `Bone = Armature.define(...)`?
- The circular dependencies of Node, GeometryNode, and Point are currently forcing me to put these classes in one file because I can't forward declare them. I'm going to do a little more research to see if there's a better way to fix this problem.

### CODE REVIEW

File: src/armature/Node.ts [hunk 1]
```diff
@@ -189,3 +214,83 @@ export class Node {
         return { ...Node.bone, transform: transformationMatrix, isShadeless: true };
     }
 }
+
+/**
+ * A derived `Node` with an additional `geometry` property.
+ */
+export class GeometryNode extends Node {
```
Comment (armcburney) [inline src/armature/Node.ts:221]: Why did you move the `GeometryNode` to the `Node` file?

**Review by armcburney — COMMENTED** [review]
I'm not a fan of having multiple classes defined in one file. I'd prefer if `GeometryNode` could be moved back to `GeometryNode.ts` and `Point` is moved to a separate file.

File: src/armature/Node.ts (hunk 1 above)
Comment (davepagurek) [reply src/armature/Node.ts:221]: Circular dependency issues. I'm hoping there's a better way but I haven't found one yet. Basically lack of forward declarations is currently causing this

Comment (armcburney) [reply src/armature/Node.ts:221]: Okay, that's fair then 👍

File: src/armature/Node.ts [hunk 2]
```diff
@@ -27,12 +27,17 @@ export class Node {
             vec3.fromValues(0, 0, 1),
             vec3.fromValues(1, 0, 0)
         ],
-        indices: [0, 1, 2, 0, 2, 3, 0, 3, 4, 5, 2, 1, 5, 3, 2, 5, 4, 3],
-        colors: Node.boneVertices.map(() => vec3.fromValues(1, 1, 1))
+        indices: [0, 1, 2, 0, 2, 3, 0, 3, 4, 0, 4, 1, 5, 2, 1, 5, 3, 2, 5, 4, 3, 5, 1, 4],
+
+        // Map x, y, z to r, g, b to give a sense of bone orientation
+        colors: Node.boneVertices.map((v: vec3) =>
+            vec3.fromValues(v[0], v[1] / 0.1 + 0.1, v[2] / 0.1 + 0.1)
```
Comment (armcburney) [inline src/armature/Node.ts:34]: What do these `0.1` constants refer to? Are they just arbitrarily defined?

Comment (davepagurek) [reply src/armature/Node.ts:34]: For each vertex, x values go between 0 and 1, y and z go between -0.1 and 0.1, so this is normalizing to get each to be between 0 and 1

**Review by armcburney — APPROVED** [review]

File: src/armature/Node.ts (hunk 1 above)
Comment (davepagurek) [reply src/armature/Node.ts:221]: I'm going to see if I can use an interface to wrap it. It's kind of messy but maybe less so than the current implementation. Will report back

Comment (davepagurek) [reply src/armature/Node.ts:221]: rip ok so I can't just use an interface because both classes need to instantiate each other as well. So I think I'm just going to leave it as is for now, if that's alright
